_drafts/2024-07-25-testing-raise.md

```diff
@@ -0,0 +1,506 @@
+---
+title: "Raise Your Tests: Testing Arrow Raise"
+date: 2024-04-10
+header:
+    image: "https://res.cloudinary.com/dkoypjlgr/image/upload/f_auto,q_auto:good,c_auto,w_1200,h_300,g_auto,fl_progressive/v1715952116/blog_cover_large_phe6ch.jpg"
+tags: [kotlin, arrow]
+excerpt: ""
+toc: true
+toc_label: "In this article"
+---
+
+_By [Riccardo Cardin](https://github.com/rcardin)_
+
+At RockTheJvm, we deeply understand the power of the Kotlin Arrow library and the Raise DSL, and we've previously shared our insights in our article on [Functional Error Handling in Kotlin, Part 3: The Raise DSL](https://blog.rockthejvm.com/functional-error-handling-in-kotlin-part-3/). Now, we're ready to delve into the crucial topic of testing applications that use the Raise DSL. To fully grasp the concepts we'll be discussing, it's essential to have a solid understanding of Kotlin, Arrow library, and the Raise DSL. We recommend revisiting our previous posts or resources if you need to refresh your knowledge. Get ready to explore the testing world of Arrow Raise with confidence and understanding.
+
+## 1. Setting up the project
+
+First things first, let's set up a Kotlin project. We can use Gradle or Maven, but for this article, we'll focus on Gradle. As always, use the `gradle init` command from a shell to create a Gradle application from scratch. We'll then add the following dependencies to the `build.gradle.kts` file, providing you with practical examples and clear explanations along the way:
+
+```kotlin
+dependencies {
+    implementation("io.arrow-kt:arrow-core:1.2.4")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.9.0-RC")
+    testImplementation("org.jetbrains.kotlin:kotlin-test-junit5")
+    testImplementation("org.junit.jupiter:junit-jupiter-engine:5.10.0")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-test:1.9.0-RC")
+    testImplementation("org.assertj:assertj-core:3.26.3")
+    testImplementation("io.kotest:kotest-runner-junit5:5.9.0")
+    testImplementation("io.kotest.extensions:kotest-assertions-arrow:1.4.0")
+    testImplementation("io.kotest.extensions:kotest-assertions-arrow-fx-coroutines:1.4.0")
+    testImplementation("org.mockito.kotlin:mockito-kotlin:5.4.0")
+}
+```
+
+Notice that we're using version `1.9.0-RC` of the coroutines library since we need support for the Kotlin compiler's version 2.0.0.
+
+We need to enable the usage of context receivers since they're still an experimental feature in Kotlin 2.0.0. We need to add the following code to the `build.gradle.kts` file to do so:
```

Comment:
I think we should prioritize the non-context-receivers style instead, while they're experimental/deprecated/uncertain, and then add an example of how the code looks better with context receivers.

Then let the reader decide which style to use in the examples that follow.